### PR TITLE
Improve test coverage

### PR DESCRIPTION
**Pull request by fzhinkin**
Added tests to cover some UTF-8-related edge cases and moved functions used primarily by tests to corresponding source sets.

### CODE REVIEW

File: core/common/test/Utf8Test.kt [hunk 1]
```diff
@@ -21,8 +21,8 @@
 
 package kotlinx.io
 
+import kotlinx.io.internal.REPLACEMENT_CHARACTER
 import kotlinx.io.internal.REPLACEMENT_CODE_POINT
-import kotlinx.io.internal.commonAsUtf8ToByteArray
 import kotlinx.io.internal.processUtf8CodePoints
 import kotlin.test.*
```
[the remark this answers is not in the document]
Comment (fzhinkin) [reply core/common/test/Utf8Test.kt:27]: > replacement behaviour is undocumented

> Although a UTF-8 conversion process is required to never consume well-formed subse- quences as part of its error handling for ill-formed subsequences, such a process is not oth- erwise constrained in how it deals with any ill-formed subsequence itself. An ill-formed subsequence consisting of more than one code unit could be treated as a single error or as multiple errors.

😢

[the remark this answers is not in the document]
Comment (qwwdfsad) [reply core/common/test/Utf8Test.kt:27]: I mean, undocumented in `kotlinx-io`.
I think we can pick a single strategy, document it and stick with it.
Right now, depending on the "ill-formness", we can replace N bytes with `[1..4]` replacement characters and it's a bit surprising. Always having a single replacement or as many characters as in an ill-formed sequence seems a bit more reasonable

[the remark this answers is not in the document]
Comment (fzhinkin) [reply core/common/test/Utf8Test.kt:27]: It's undocumented in the standard, either. :)

And, IIRC, we also have the same discrepancy in stdlib, where native and JVM implementations of byte-array-to-string conversion.

File: core/common/test/Utf8Test.kt [hunk 2]
```diff
@@ -21,8 +21,8 @@
 
 package kotlinx.io
 
+import kotlinx.io.internal.REPLACEMENT_CHARACTER
 import kotlinx.io.internal.REPLACEMENT_CODE_POINT
-import kotlinx.io.internal.commonAsUtf8ToByteArray
```
[the remark this answers is not in the document]
Comment (fzhinkin) [reply core/common/test/Utf8Test.kt:25]: Calling String's ctor seems to be slower (by about 10-15% when it comes to utf8 strings) compared to what we do now.

File: core/common/test/Utf8Test.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (fzhinkin) [reply core/common/test/Utf8Test.kt:27]: Speaking of examples you initially gave, it all seems more or less consistent:
- `0xf0 0x89 0x89 <EOF>` - the sequence has a valid prefix, but terminates abruplty -> we replaced the whole sequence with `\ufffd`;
- `0xf0 0x89 0x89 0x89 <EOF>` - the sequence is valid, but encoded value lies outside the valid range -> we replaced the whole sequence with `\ufffd`;
- `0xf0 0xf0 0xf0 <EOF>` - we don't consider `0xf0 0xf0` as a sequence as the second `0xf0` is not a continuation CP, thus the first invalid sequence consists of a single byte, we replaced, and did the same for other two single-byte sequences.

However, looking at UTF-8's definitions of well- and ill-formed cp-sequences (D84-D86) gives an impression that `0xf0 0xf0 0xf0` should be treated as a single ill-formed sequence as none of its bytes overlaps with a minimal well-formed subsequence. 🤔

[the remark this answers is not in the document]
Comment (fzhinkin) [reply core/common/test/Utf8Test.kt:27]: But current behavior is a conforming one:

> An ill-formed subsequence consisting of more than one code unit could be treated as a single error or as multiple errors.

File: core/common/test/Utf8Test.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (qwwdfsad) [reply core/common/test/Utf8Test.kt:25]: This is really unexpected though 🥲

File: core/common/test/Utf8Test.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (fzhinkin) [reply core/common/test/Utf8Test.kt:27]: @qwwdfsad, I filed https://github.com/Kotlin/kotlinx-io/issues/301 to track it

**Review by qwwdfsad — APPROVED** [review]
